Display structured logs in admin

fabrixxm opened the pull request:
This PR adds a structured display of friendica logs in admin.

**wip** waiting for next dev cycle, open to discussions if all of this it's worth, or if could be done in better ways.

This patch tries to parse log lines and to display info in a table. Additional JSON data is parsed and displayed clicking on a row.

Logs are read from current log file. The log format is not changed, to keep compatibility with current logs,  although this complicates a bit the parsing.

`Friendica\Module\Admin\Logs\View` uses `Friendica\Model\Log\ParsedLogIterator` to lazy-load log lines;

`Friendica\Model\Log\ParsedLogIterator` load log file using `\Friendica\Util\ReversedFileReader`  and returns one `\Friendica\Objec\Log\ParsedLog` per `next()`;

`\Friendica\Util\ReversedFileReader` is an iterator which reurns lines from file starting from the end going backward;

`\Friendica\Objec\Log\ParsedLog` contains the logic to parse a single log line and holds data to be shown in template.

![immagine](https://user-images.githubusercontent.com/309361/119325782-87cb9000-bc81-11eb-97df-bc84ae1dccd5.png)

![immagine](https://user-images.githubusercontent.com/309361/119326013-c3665a00-bc81-11eb-84db-2eb36d946ada.png)

- [x] File reading and line parsing is handled in iterators, to avoid to keep  too much data in memory.
- [x] Log file is read backward to display log events newest first.
- [x] Add search and filter.
~~Add "tail" functionality.~~
- [x] Use DI
- [x] Add tests
- [x] Make code coding-standards compliant :)

fabrixxm committed:
Display structured logs in admin

Tries to parse log lines and to display info in a table.
Additional JSON data is parsed and displayed clicking on a row.

File reading and line parsing is handled in iterators, to avoid to keep
too much data in memory.
Search and filter should be trivial to add.
Log file is read backward to display log events newest first.
A "tail" functionality should be easy to implement.

fabrixxm committed:
Advanced log view: Frio version

Show log event details in a modal dialog

fabrixxm commented:
And: this PR right now is toward `2021.06-rc` branch just because is currently the most advanced branch.
It should go in `develop` when is ready.

MrPetovan commented:
I converted this PR to Draft to prevent someone from merging it too soon. You'll be able to convert it back to a regular PR when you deem it ready.

MrPetovan commented:
Misclick.

fabrixxm committed:
Add search and filter to log view

fabrixxm committed:
Fix code formatting

fabrixxm committed:
logs view: update frio template

add search and filter columns support.
add "prev/next" buttons to details popup

tobiasd commented:
> It should go in develop when is ready.

I've changed it back towards `develop`.

fabrixxm committed:
Add tests for ParsedLog and fix parsing

fabrixxm committed:
Add DI to ParsedLogIterator, replace constructors with fluent api

fabrixxm committed:
Add tests for ParsedLogIterator

fabrixxm committed:
Fix code standard

fabrixxm committed:
Fix `ParsedLogIteratorTest::setUp()`

MrPetovan commented:
Just a thousand lines of code to review, no biggie 😅

MrPetovan commented on src/Model/Log/ParsedLogIterator.php:
```diff
@@ -0,0 +1,189 @@
+	{
+		$parsed = $this->read();
+
+		// if read() has not retuned none and
+		// the line don't match filters or search
+		// 	read the next line
```
This comment is an exact duplicate of the code as it is written. Its presence is extraneous, which is a good thing, you wrote understandable code!
```suggestion
```

MrPetovan commented on src/Model/Log/ParsedLogIterator.php:
```diff
@@ -0,0 +1,189 @@
+		return new ParsedLog($this->reader->key(), $line);
+	}
+
+	public function next()
```
This whole method needs explaining, especially its relationship with search and filters and why `$this->value` can end up being assigned `null` as part of the while loop break condition, and why it doesn't return any value.

fabrixxm replied:
I've added docstring with a quick explanation and a remaind to `Iterator` interface documentation

MrPetovan commented on src/Object/Log/ParsedLog.php:
```diff
@@ -0,0 +1,147 @@
+/**
+ * Parse a log line and offer some utility methods
+ */
+class ParsedLog
```
Since it's about a single line, I believe it should be explicit in the name.
```suggestion
class ParsedLogLine
```
Please don't forget to update all the references to this class.

MrPetovan commented on src/Object/Log/ParsedLog.php:
```diff
@@ -0,0 +1,147 @@
+		} catch (\JsonException $e) {
+			// try to find next { in $str and move string before to 'message'
+
+			$pos = strpos($this->data, '{', 1);
```
What happens when `$pos` becomes `false`?

MrPetovan commented on src/Object/Log/ParsedLog.php:
```diff
@@ -0,0 +1,147 @@
+	 * This method try to parse the found json and if it fails, search for next '{'
+	 * in json data and retry
+	 */
+	private function try_fix_json()
```
Please use `camelCase` for method names as you did in `ParsedLogIterator`
```suggestion
	private function tryFixJson()
```
Please don't forget to update all the references to this method.

MrPetovan commented on src/Object/Log/ParsedLog.php:
```diff
@@ -0,0 +1,147 @@
+	 *
+	 * @return array
+	 */
+	public function get_data()
```
Please use `camelCase` for method names as you did in `ParsedLogIterator`
```suggestion
	public function getData()
```
Please don't forget to update all the references to this method.

MrPetovan commented on src/Object/Log/ParsedLog.php:
```diff
@@ -0,0 +1,147 @@
+	 *
+	 * @return array
+	 */
+	public function get_source()
```
Please use `camelCase` for method names as you did in `ParsedLogIterator`
```suggestion
	public function getSource()
```
Please don't forget to update all the references to this method.

MrPetovan commented on tests/src/Model/Log/ParsedLogIteratorTest.php:
```diff
@@ -0,0 +1,153 @@
+
+	protected function setUp(): void
+	{
+		$logfile = dirname(__DIR__) . DIRECTORY_SEPARATOR .
```
The use of `DIRECTORY_SEPARATOR` is laudable but in practice verbose and useless, all relevant platforms understand `/`, Windows included.

MrPetovan commented on view/templates/admin/logs/view.tpl:
```diff
@@ -1,6 +1,74 @@
 	<h1>{{$title}} - {{$page}}</h1>
-	
+
 	<h3>{{$logname}}</h3>
```
No `<h2>` ?

MrPetovan commented on view/templates/admin/logs/view.tpl:
```diff
@@ -1,6 +1,74 @@
+				</thead>
+				<tbody>
+					{{foreach $data as $row}}
+						<tr id="ev-{{$row->id}}" onClick="log_show_details('ev-{{$row->id}}')">
```
Please add a visual and text cues that this element can be clicked, through a cursor change and an `alt` attribute.

fabrixxm replied:
as per MDN `alt` is an attribute for `<img>` tags only.
what about `aria-label`?
with `role="button" tabIndex="0" aria-label="View details" aria-haspopup="true" aria-expanded="false"`, css `cursor:pointer` and js code to toggle `aria-expanded` value and react to "space" and "enter" keypress, should be correctly accessible..

MrPetovan replied:
This sounds excellent, thank you for looking into it.

MrPetovan commented on view/theme/frio/templates/admin/logs/view.tpl:
```diff
@@ -0,0 +1,136 @@
+<div id="adminpage">
+	<h1>{{$title}} - {{$page}}</h1>
+
+	<h3>{{$logname}}</h3>
```
No `<h2>` ?

MrPetovan commented on view/theme/frio/templates/admin/logs/view.tpl:
```diff
@@ -0,0 +1,136 @@
+			</thead>
+			<tbody>
+				{{foreach $data as $row}}
+				<tr id="ev-{{$row->id}}" class="log-event"
```
Please add a visual and text cues that this element can be clicked, through a cursor change and an `alt` attribute.

fabrixxm committed:
Fix review points

- Fix headers hierarchy
- Improve accessibility:
 	- set mouse pointer
	- make rows focusable
	- open on key press
	- add tooltip with "title"
	- add role and aria attributes
- Rename `ParsedLog` to `ParsedLogLine`
- Add docs to `ReversedFileReader`'s implementation of `Iterator`'s methods
- Add docs to `ParsedLogIterator`'s implementation of `Iterator`'s methods
- Remove unnecessary comment
- Add more test for parsing log lines and fix some edge cases
- Fix function name in snake-case to camelCase
- Remove `DIRECTORY_SEPARATOR`

fabrixxm committed:
Fix code standard

fabrixxm commented:
> Just a thousand lines of code to review, no biggie 😅

Yes, I know. Is huge. I'm sorry 😊
